### PR TITLE
Add support for Sifive PWM.

### PR DESCRIPTION
**Pull request by NandkumarJoshi**

### CODE REVIEW

File: metal_header/sifive_pwm0.c++ [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+                        "(struct __metal_driver_sifive_gpio0 *)&__metal_dt_" +
+                        m.handle();
+                    pinmux_dest = dest;
+                    pinmux_source = source;
+                  });
+
+    /* Comparator count and width */
+    uint32_t cmpwidth = 0;
+    uint32_t ncmp = 0;
+    n.maybe_tuple("ncmp_cmpwidth", tuple_t<uint32_t, uint32_t>(), [&]() {},
```
Comment (bsousi5) [inline metal_header/sifive_pwm0.c++]: The string should be "sifive,comparator-widthbits". Also check for existence, if not available default to 16.
Similarly, check for existence if "sifive,ncomparators". If not available, default to 1.

**Comment by NandkumarJoshi** [comment]
> You also need bare_metal header generation for different PWM-CMP reg width
> See https://github.com/sifive/federation/blob/master/software/tests/pwm/pwm.c for reference.

Max PWM cmp_width is 16bit, which makes pwmcount reg max 32bit wide. It will not affect register offsets.

File: metal_header/sifive_pwm0.c++ (hunk 1 above)
Comment (NandkumarJoshi) [reply metal_header/sifive_pwm0.c++]: String updated, existence check added. Default values are set.

File: metal_header/sifive_pwm0.c++ [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+  dtb.match(std::regex(compat_string), [&](node n) {
+    /* Clock driving the pwm peripheral */
+    std::string clock_value = "NULL";
+    n.maybe_tuple("clocks", tuple_t<node>(), [&]() {},
+                  [&](node m) {
+                    clock_value = "(struct metal_clock *)&__metal_dt_" +
+                                  m.handle() + ".clock";
+                  });
+
+    /* Pinmux */
```
[the remark this answers is not in the document]
Comment (NandkumarJoshi) [reply metal_header/sifive_pwm0.c++]: Pin mux is present in SiFive FE310-G002 Manual v19p04, page 78, table 53.
Will be don't care, if any implementation lacks a pin-mux. Default values are set to "0".

**Pull request closed without merging**